### ISSUE TITLE
Question regarding detach argument 

### ISSUE THREAD

**Issue by MushroomHunting** [chijames/GST#2]
Hi there,

First of wanted to say thanks for the awesome work!

I have a question regarding the gst and rao gumbel samplers. In the first line of each there is this line of code

`logits_cpy = logits.detach() if detach else logits
`

with `detach=True` being the default. What is the significance of setting this to True vs False?

In my own testing I have tried a trivial example sampling from a bunch of Bernoulli variables with the loss simply being to require the sum of all the first classes of N individual trials to sum to some constant C. When `detach=True`, the training loss is quite noisy (even noisier than just gumbel straight through) but eventually converges. When `detach = False`, the training loss converges orders of magnitude faster. Although this is just for a trivial example I'm curious what the implications are for more general models. Why is True the default option? If this is discussed in the main paper i'd be grateful to be directed to a specific section

my intialisation looks like this

```
N = 100000
C = 50000
probs_pos = np.random.uniform(0.1, 0.9, (1, N, 1))
probs_neg = 1 - probs_pos 
probs = np.concatenate([probs_pos, probs_neg], axis=-1)
logits = torch.Tensor(np.log(probs))

```
the sampler module looks like
```
class SimpleModel(nn.Module):
    def __init__(self,
                 logits: torch.Tensor,
                 detach_copy: bool = True,
                 dtype_torch=torch.float32) -> None:
        super(SimpleModel, self).__init__()
        self.logits_init = logits
        self.detach_copy = detach_copy
        self.dtype_torch = dtype_torch
        self.logits = nn.Parameter(self.logits_init, requires_grad=True)

    def forward(self):
        samples = gst_mover(logits=self.logits,
                      temperature=self.temp,
                      gap=self.gap,
                      hard=True,
                      mask=None,
                      detach=self.detach_copy)
        return samples
```
and the training loop

```
detach_copy = True
# detach_copy = False
...
model = SimpleModel(logits, detach_copy=detach_copy)
...
for epoch in training_loop_iterator:
    optimizer.zero_grad()  
    pred = model()  # simply return the gumbel sample
    pred = pred.squeeze()
    vals = pred[:, 0]

    loss = simple_loss(vals, total=C)  
    loss.backward()
    optimizer.step()
```
with loss defined as

```
def simple_loss(samples, total: float):
    loss = (total - samples.sum()) ** 2
    return loss
```

cheers!

**Comment by chijames** [comment]
Hi,

Thank you for trying out GST!

The reason of setting detach to True is discussed in section 3.2 of our [paper](https://proceedings.mlr.press/v162/fan22a/fan22a.pdf). In short, backpropagating gradients through both the logits and perturbation noises causes the training process to be unstable in our experiments.

Can you try using GST-pi (with detach=True) instead of GST-1.0? You can do so by calling gst_mover(logits, gap=‘p’). There is another [paper](https://arxiv.org/pdf/2304.08612.pdf) that also reports inferior performance of gst-1.0 on certain tasks, while gst-pi performs much better on them.

Thanks!

**Comment by MushroomHunting** [comment]
Hi chijames

thanks so much for your response - this did make a difference! Thanks also for pointing out the section in the paper :+1:
